File: pkg/dev_compiler/test/codegen_expected/corelib/regexp/capture-3_test.js
dart_library.library('corelib/regexp/capture-3_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__capture$453_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const capture$453_test = Object.create(null);
  const v8_regexp_utils = Object.create(null);
  let VoidTovoid = () => (VoidTovoid = dart.constFn(dart.functionType(dart.void, [])))();
  let ListOfString = () => (ListOfString = dart.constFn(core.List$(core.String)))();
  let dynamicTodynamic = () => (dynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic])))();
  let VoidTovoid$ = () => (VoidTovoid$ = dart.constFn(dart.definiteFunctionType(dart.void, [])))();
  let dynamicAnddynamic__Tovoid = () => (dynamicAnddynamic__Tovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic, dart.dynamic], [core.String])))();
  let dynamic__Tovoid = () => (dynamic__Tovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic], [core.String])))();
  let dynamic__Tovoid$ = () => (dynamic__Tovoid$ = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic], [core.num])))();
  let dynamicAnddynamicAndnumTovoid = () => (dynamicAnddynamicAndnumTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic, dart.dynamic, core.num])))();
  let dynamicTovoid = () => (dynamicTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic])))();
  let StringAndRegExpToMatch = () => (StringAndRegExpToMatch = dart.constFn(dart.definiteFunctionType(core.Match, [core.String, core.RegExp])))();
  let MatchToString = () => (MatchToString = dart.constFn(dart.definiteFunctionType(core.String, [core.Match])))();
  let StringAndRegExpToListOfString = () => (StringAndRegExpToListOfString = dart.constFn(dart.definiteFunctionType(ListOfString(), [core.String, core.RegExp])))();
  capture$453_test.main = function() {
    function oneMatch(re) {
      v8_regexp_utils.assertEquals("acd", "abcd"[dartx.replaceAll](core.Pattern._check(re), ""));
    }
    dart.fn(oneMatch, dynamicTodynamic());
    oneMatch(core.RegExp.new("b"));
    oneMatch(core.RegExp.new("b"));
    v8_regexp_utils.assertEquals("acdacd", "abcdabcd"[dartx.replaceAll](core.RegExp.new("b"), ""));
    function captureMatch(re) {
      let match = v8_regexp_utils.firstMatch("abcd", core.RegExp._check(re));
      v8_regexp_utils.assertEquals("b", match.group(1));
      v8_regexp_utils.assertEquals("c", match.group(2));
    }
    dart.fn(captureMatch, dynamicTodynamic());
    captureMatch(core.RegExp.new("(b)(c)"));
    captureMatch(core.RegExp.new("(b)(c)"));
    let a = "foo bar baz"[dartx.replaceAll](core.RegExp.new("^|bar"), "");
    v8_regexp_utils.assertEquals("foo  baz", a);
    a = "foo bar baz"[dartx.replaceAll](core.RegExp.new("^|bar"), "*");
    v8_regexp_utils.assertEquals("*foo * baz", a);
    function NoHang(re) {
      v8_regexp_utils.firstMatch("This is an ASCII string that could take forever", core.RegExp._check(re));
    }
    dart.fn(NoHang, dynamicTodynamic());
    NoHang(core.RegExp.new("(((.*)*)*x)Ā"));
    NoHang(core.RegExp.new("(((.*)*)*Ā)foo"));
    NoHang(core.RegExp.new("Ā(((.*)*)*x)"));
    NoHang(core.RegExp.new("(((.*)*)*x)Ā"));
    NoHang(core.RegExp.new("[ćăĀ](((.*)*)*x)"));
    NoHang(core.RegExp.new("(((.*)*)*x)[ćăĀ]"));
    NoHang(core.RegExp.new("[^\\x00-\\xff](((.*)*)*x)"));
    NoHang(core.RegExp.new("(((.*)*)*x)[^\\x00-\\xff]"));
    NoHang(core.RegExp.new("(?!(((.*)*)*x)Ā)foo"));
    NoHang(core.RegExp.new("(?!(((.*)*)*x))Ā"));
    NoHang(core.RegExp.new("(?=(((.*)*)*x)Ā)foo"));
    NoHang(core.RegExp.new("(?=(((.*)*)*x))Ā"));
    NoHang(core.RegExp.new("(?=Ā)(((.*)*)*x)"));
    NoHang(core.RegExp.new("(æ|ø|Ā)(((.*)*)*x)"));
    NoHang(core.RegExp.new("(a|b|(((.*)*)*x))Ā"));
    NoHang(core.RegExp.new("(a|(((.*)*)*x)ă|(((.*)*)*x)Ā)"));
    let s = "Don't prune based on a repetition of length 0";
    v8_regexp_utils.assertEquals(null, v8_regexp_utils.firstMatch(s, core.RegExp.new("å{1,1}prune")));
    v8_regexp_utils.assertEquals("prune", v8_regexp_utils.firstMatch(s, core.RegExp.new("å{0,0}prune")).get(0));
    let regex6 = core.RegExp.new("a*\\u0100*\\w");
    let input0 = "a";
    regex6.firstMatch(input0);
    let re = "Ā*\\w";
    for (let i = 0; i < 200; i++)
      re = "a*" + re;
    let regex7 = core.RegExp.new(re);
    regex7.firstMatch(input0);
    let regex8 = core.RegExp.new(re, {caseSensitive: false});
    regex8.firstMatch(input0);
    re = "[Ā]*\\w";
    for (let i = 0; i < 200; i++)
      re = "a*" + re;
    let regex9 = core.RegExp.new(re);
    regex9.firstMatch(input0);
    let regex10 = core.RegExp.new(re, {caseSensitive: false});
    regex10.firstMatch(input0);
    let regex11 = core.RegExp.new("^(?:[^\\u0000-\\u0080]|[0-9a-z?,.!&\\s#()])+$", {caseSensitive: false});
    regex11.firstMatch(input0);
    let regex12 = core.RegExp.new("u(\\xf0{8}?\\D*?|( ? !)$h??(|)*?(||)+?\\6((?:\\W\\B|--\\d-*-|)?$){0, }?|^Y( ? !1)\\d+)+a");
    regex12.firstMatch("");
  };
  dart.fn(capture$453_test.main, VoidTovoid$());
  v8_regexp_utils.assertEquals = function(actual, expected, message) {
    if (message === void 0) message = null;
    expect$.Expect.equals(actual, expected, message);
  };
  dart.fn(v8_regexp_utils.assertEquals, dynamicAnddynamic__Tovoid());
  v8_regexp_utils.assertTrue = function(actual, message) {
    if (message === void 0) message = null;
    expect$.Expect.isTrue(actual, message);
  };
  dart.fn(v8_regexp_utils.assertTrue, dynamic__Tovoid());
  v8_regexp_utils.assertFalse = function(actual, message) {
    if (message === void 0) message = null;
    expect$.Expect.isFalse(actual, message);
  };
  dart.fn(v8_regexp_utils.assertFalse, dynamic__Tovoid());
  v8_regexp_utils.assertThrows = function(fn, testid) {
    if (testid === void 0) testid = null;
    expect$.Expect.throws(VoidTovoid()._check(fn), null, dart.str`Test ${testid}`);
  };
  dart.fn(v8_regexp_utils.assertThrows, dynamic__Tovoid$());
  v8_regexp_utils.assertNull = function(actual, testid) {
    if (testid === void 0) testid = null;
    expect$.Expect.isNull(actual, dart.str`Test ${testid}`);
  };
  dart.fn(v8_regexp_utils.assertNull, dynamic__Tovoid$());
  v8_regexp_utils.assertToStringEquals = function(str, match, testid) {
    let actual = [];
    for (let i = 0; i <= dart.notNull(core.num._check(dart.dload(match, 'groupCount'))); i++) {
      let g = dart.dsend(match, 'group', i);
      actual[dartx.add](g == null ? "" : g);
    }
    expect$.Expect.equals(str, actual[dartx.join](","), dart.str`Test ${testid}`);
  };
  dart.fn(v8_regexp_utils.assertToStringEquals, dynamicAnddynamicAndnumTovoid());
  v8_regexp_utils.shouldBeTrue = function(actual) {
    expect$.Expect.isTrue(actual);
  };
  dart.fn(v8_regexp_utils.shouldBeTrue, dynamicTovoid());
  v8_regexp_utils.shouldBeFalse = function(actual) {
    expect$.Expect.isFalse(actual);
  };
  dart.fn(v8_regexp_utils.shouldBeFalse, dynamicTovoid());
  v8_regexp_utils.shouldBeNull = function(actual) {
    expect$.Expect.isNull(actual);
  };
  dart.fn(v8_regexp_utils.shouldBeNull, dynamicTovoid());
  v8_regexp_utils.shouldBe = function(actual, expected, message) {
    if (message === void 0) message = null;
    if (expected == null) {
      expect$.Expect.isNull(actual, message);
    } else {
      expect$.Expect.equals(dart.dload(expected, 'length'), dart.dsend(dart.dload(actual, 'groupCount'), '+', 1));
      for (let i = 0; i <= dart.notNull(core.num._check(dart.dload(actual, 'groupCount'))); i++) {
        expect$.Expect.equals(dart.dindex(expected, i), dart.dsend(actual, 'group', i), message);
      }
    }
  };
  dart.fn(v8_regexp_utils.shouldBe, dynamicAnddynamic__Tovoid());
  v8_regexp_utils.firstMatch = function(str, pattern) {
    return pattern.firstMatch(str);
  };
  dart.fn(v8_regexp_utils.firstMatch, StringAndRegExpToMatch());
  v8_regexp_utils.allStringMatches = function(str, pattern) {
    return pattern.allMatches(str)[dartx.map](core.String)(dart.fn(m => m.group(0), MatchToString()))[dartx.toList]();
  };
  dart.fn(v8_regexp_utils.allStringMatches, StringAndRegExpToListOfString());
  v8_regexp_utils.description = function(str) {
  };
  dart.fn(v8_regexp_utils.description, dynamicTovoid());
  // Exports:
  exports.capture$453_test = capture$453_test;
  exports.v8_regexp_utils = v8_regexp_utils;
});

File: tests/corelib/regexp/capture-3_test.dart
// Copyright (c) 2014, the Dart project authors. All rights reserved.
// Copyright 2009 the V8 project authors. All rights reserved.
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
//       copyright notice, this list of conditions and the following
//       disclaimer in the documentation and/or other materials provided
//       with the distribution.
//     * Neither the name of Google Inc. nor the names of its
//       contributors may be used to endorse or promote products derived
//       from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import 'v8_regexp_utils.dart';
import 'package:expect/expect.dart';

void main() {
  dynamic oneMatch(re) {
    assertEquals("acd", "abcd".replaceAll(re, ""));
  }

  oneMatch(new RegExp(r"b"));
  oneMatch(new RegExp(r"b"));

  assertEquals("acdacd", "abcdabcd".replaceAll(new RegExp(r"b"), ""));

  dynamic captureMatch(re) {
    var match = firstMatch("abcd", re);
    assertEquals("b", match.group(1));
    assertEquals("c", match.group(2));
  }

  captureMatch(new RegExp(r"(b)(c)"));
  captureMatch(new RegExp(r"(b)(c)"));

  // A test that initially does a zero width match, but later does a non-zero
  // width match.
  var a = "foo bar baz".replaceAll(new RegExp(r"^|bar"), "");
  assertEquals("foo  baz", a);

  a = "foo bar baz".replaceAll(new RegExp(r"^|bar"), "*");
  assertEquals("*foo * baz", a);

  // We test FilterASCII using regexps that will backtrack forever.  Since
  // a regexp with a non-ASCII character in it can never match an ASCII
  // string we can test that the relevant node is removed by verifying that
  // there is no hang.
  dynamic NoHang(re) {
    firstMatch("This is an ASCII string that could take forever", re);
  }

  NoHang(new RegExp(
      r"(((.*)*)*x)Ā")); // Continuation after loop is filtered, so is loop.
  NoHang(new RegExp(r"(((.*)*)*Ā)foo")); // Body of loop filtered.
  NoHang(new RegExp(
      r"Ā(((.*)*)*x)")); // Everything after a filtered character is filtered.
  NoHang(new RegExp(
      r"(((.*)*)*x)Ā")); // Everything before a filtered character is filtered.
  NoHang(new RegExp(
      r"[ćăĀ](((.*)*)*x)")); // Everything after a filtered class is filtered.
  NoHang(new RegExp(
      r"(((.*)*)*x)[ćăĀ]")); // Everything before a filtered class is filtered.
  NoHang(new RegExp(r"[^\x00-\xff](((.*)*)*x)")); // After negated class.
  NoHang(new RegExp(r"(((.*)*)*x)[^\x00-\xff]")); // Before negated class.
  NoHang(new RegExp(r"(?!(((.*)*)*x)Ā)foo")); // Negative lookahead is filtered.
  NoHang(new RegExp(
      r"(?!(((.*)*)*x))Ā")); // Continuation branch of negative lookahead.
  NoHang(new RegExp(r"(?=(((.*)*)*x)Ā)foo")); // Positive lookahead is filtered.
  NoHang(new RegExp(
      r"(?=(((.*)*)*x))Ā")); // Continuation branch of positive lookahead.
  NoHang(new RegExp(
      r"(?=Ā)(((.*)*)*x)")); // Positive lookahead also prunes continuation.
  NoHang(new RegExp(
      r"(æ|ø|Ā)(((.*)*)*x)")); // All branches of alternation are filtered.
  NoHang(new RegExp(r"(a|b|(((.*)*)*x))Ā")); // 1 out of 3 branches pruned.
  NoHang(new RegExp(
      r"(a|(((.*)*)*x)ă|(((.*)*)*x)Ā)")); // 2 out of 3 branches pruned.

  var s = "Don't prune based on a repetition of length 0";
  assertEquals(null, firstMatch(s, new RegExp(r"å{1,1}prune")));
  assertEquals("prune", (firstMatch(s, new RegExp(r"å{0,0}prune"))[0]));

  // Some very deep regexps where FilterASCII gives up in order not to make the
  // stack overflow.
  var regex6 = new RegExp(r"a*\u0100*\w");
  var input0 = "a";
  regex6.firstMatch(input0);

  var re = "\u0100*\\w";

  for (var i = 0; i < 200; i++) re = "a*" + re;

  var regex7 = new RegExp(re);
  regex7.firstMatch(input0);

  var regex8 = new RegExp(re, caseSensitive: false);
  regex8.firstMatch(input0);

  re = "[\u0100]*\\w";
  for (var i = 0; i < 200; i++) re = "a*" + re;

  var regex9 = new RegExp(re);
  regex9.firstMatch(input0);

  var regex10 = new RegExp(re, caseSensitive: false);
  regex10.firstMatch(input0);

  var regex11 = new RegExp(r"^(?:[^\u0000-\u0080]|[0-9a-z?,.!&\s#()])+$",
      caseSensitive: false);
  regex11.firstMatch(input0);

  var regex12 = new RegExp(
      r"u(\xf0{8}?\D*?|( ? !)$h??(|)*?(||)+?\6((?:\W\B|--\d-*-|)?$){0, }?|^Y( ? !1)\d+)+a");
  regex12.firstMatch("");
}
